xds/internal/xdsclient/authority.go

```diff
@@ -86,89 +98,132 @@ type authority struct {
+// This method is called by the xDS client implementation (on all interested
+// authorities) when a stream error is reported by the xdsChannel.
+//
+// Errors of type xdsresource.ErrTypeStreamFailedAfterRecv are ignored.
+func (a *authority) adsStreamFailure(serverConfig *bootstrap.ServerConfig, err error) {
+	a.serializer.TrySchedule(func(context.Context) {
+		a.handleADSStreamFailure(serverConfig, err)
+	})
 }
 
-func (a *authority) handleResourceUpdate(resourceUpdate transport.ResourceUpdate, onDone func()) error {
-	rType := a.resourceTypeGetter(resourceUpdate.URL)
-	if rType == nil {
-		return xdsresource.NewErrorf(xdsresource.ErrorTypeResourceTypeUnsupported, "Resource URL %v unknown in response from server", resourceUpdate.URL)
+// Handles ADS stream failure by invoking watch callbacks and triggering
+// fallback if the associated conditions are met.
+//
+// Only executed in the context of a serializer callback.
+func (a *authority) handleADSStreamFailure(serverConfig *bootstrap.ServerConfig, err error) {
+	if a.logger.V(2) {
+		a.logger.Infof("Connection to server %s failed with error: %v", serverConfig, err)
 	}
 
-	opts := &xdsresource.DecodeOptions{
-		BootstrapConfig: a.bootstrapCfg,
-		ServerConfig:    a.serverCfg,
+	// We do not consider it an error if the ADS stream was closed after having
+	// received a response on the stream. This is because there are legitimate
+	// reasons why the server may need to close the stream during normal
+	// operations, such as needing to rebalance load or the underlying
+	// connection hitting its max connection age limit. See gRFC A57 for more
+	// details.
+	if xdsresource.ErrType(err) == xdsresource.ErrTypeStreamFailedAfterRecv {
+		a.logger.Warningf("Watchers not notified since ADS stream failed after having received at least one response: %v", err)
+		return
+	}
+
+	// Propagate the connection error from the transport layer to all watchers.
+	for _, rType := range a.resources {
+		for _, state := range rType {
+			for watcher := range state.watchers {
+				watcher := watcher
+				a.watcherCallbackSerializer.TrySchedule(func(context.Context) {
+					watcher.OnError(xdsresource.NewErrorf(xdsresource.ErrorTypeConnection, "xds: error received from xDS stream: %v", err), func() {})
+				})
+			}
+		}
 	}
-	updates, md, err := decodeAllResources(opts, rType, resourceUpdate)
-	a.updateResourceStateAndScheduleCallbacks(rType, updates, md, onDone)
-	return err
+
+	// TODO(easwars-fallback): Trigger fallback here if conditions for fallback
+	// are met.
 }
 
-func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Type, updates map[string]resourceDataErrTuple, md xdsresource.UpdateMetadata, onDone func()) {
-	a.resourcesMu.Lock()
-	defer a.resourcesMu.Unlock()
+// adsResourceUpdate is called to notify the authority about a resource update
+// received on the ADS stream. It processes the update, updating the resource
+// cache and notifying any registered watchers of the update.
```

Comment:
Nit: all these functions called by the client just potentially try schedule the handle function. Move this comment to the function below handleADSResourceUpdate.